src/ai/backend/manager/server.py

```diff
@@ -490,6 +491,113 @@ async def monitoring_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
         await ectx.cleanup()
 
 
+@actxmgr
+async def hanging_sessions_scanner_ctx(root_ctx: RootContext) -> AsyncIterator[None]:
+    from contextlib import suppress
+    from datetime import timedelta
+    from typing import TYPE_CHECKING
+
+    import sqlalchemy as sa
+    from dateutil.tz import tzutc
+    from sqlalchemy.orm import load_only, noload
+
+    from .models.session import SessionStatus
+
+    if TYPE_CHECKING:
+        from .models.utils import ExtendedAsyncSAEngine
+
+    async def _fetch_hanging_sessions(
+        db: ExtendedAsyncSAEngine,
+        status: SessionStatus,
+        threshold: timedelta,
+    ) -> tuple[SessionRow, ...]:
+        query = (
+            sa.select(SessionRow)
+            .where(SessionRow.status == status)
+            .where(
+                (
+                    datetime.now(tz=tzutc())
+                    - SessionRow.status_history[status.name].astext.cast(
+                        sa.types.DateTime(timezone=True)
+                    )
+                )
+                > threshold
+            )
+            .options(
+                noload("*"),
+                load_only(SessionRow.id, SessionRow.name, SessionRow.status, SessionRow.access_key),
+            )
+        )
+        async with db.begin_readonly() as conn:
+            result = await conn.execute(query)
+            return result.fetchall()
+
+    async def _force_terminate_hanging_sessions(
+        status: SessionStatus,
+        threshold: timedelta,
+        reason: KernelLifecycleEventReason = KernelLifecycleEventReason.HANGTIME_EXCEEDED,
+    ) -> None:
+        while True:
+            sessions = await _fetch_hanging_sessions(
+                root_ctx.db, status=status, threshold=threshold
+            )
+            log.debug(f"{len(sessions)} {status} sessions found.")
+
+            _ = await asyncio.gather(
+                *[
+                    asyncio.create_task(
+                        root_ctx.registry.destroy_session(session, forced=True, reason=reason),
+                    )
+                    for session in sessions
+                ]
+            )
+
+            await asyncio.sleep(threshold.seconds)
```

Comment:
The check interval should be shorter than the timeout threshold, but it shouldn't be too frequent to avoid database overheads.
There is no golden rule here, but I think at least we need to use an interval less than *the half* of the threshold to achieve a sufficient frequency. (i.e., The wavelength to get a desired resolution should be the half of the target resolution. In the same way, usually we set the heartbeat timeout as a double of heartbeat interval.)

Reply:
Resolved: 395ea3b7d946e743422a630b71c413707cb786d9, 9604d3c9a12970745df5c558d9c05064a4541eb2, 76c311cb70c00ea3b5cbd225e6e65af2b390f4e1
I set the interval to 40% of the given threshold.